### PR TITLE
Dynamic scanning model

### PR DESCRIPTION
**Pull request by Col-E**
Example of format for partial matching of https://stackoverflow.com/questions/62289/read-write-to-windows-registry-using-java
```json
{
  "archetype": {
    "level": "MAXIMUM",
    "identifier": "Windows Registry",
    "description": "Indicators of working with Windows Registry"
  },
  "code-patterns": {
    "invokeRootUser": [{ "op": "STARTS_WITH invoke", "args": "STARTS_WITH java/util/prefs/Preferences.userRoot()" }],
    "invokeRootSysm": [{ "op": "STARTS_WITH invoke", "args": "STARTS_WITH java/util/prefs/Preferences.systemRoot()" }]
  },
  "code-behaviors": {
    "lookupMethod": {
      "location": { "class": "EQUALS java/lang/Class", "mname": "EQUALS getDeclaredMethod" },
      "condition": { "ANY": [
        { "index": 1, "match": "EQUALS WindowsRegOpenKey" },
        { "index": 1, "match": "EQUALS WindowsRegCloseKey" },
        { "index": 1, "match": "EQUALS WindowsRegQueryValueEx" },
        { "index": 1, "match": "EQUALS WindowsRegEnumValue" },
        { "index": 1, "match": "EQUALS WindowsRegEnumKeyEx" },
        { "index": 1, "match": "EQUALS WindowsRegSetValueEx" },
        { "index": 1, "match": "EQUALS WindowsRegDeleteValue" },
        { "index": 1, "match": "EQUALS WindowsRegDeleteKey" }
      ]}
    }
  }
}
```

PR includes:
- Models to support dynamic signatures
- Basic plugging into the existing SSVM pipeline

PR does _not_ include:
- Implementations for `EntryPointDiscovery` or `CoverageEntryPointSupplier` which are required for the dynamic scanner to actually work
    - That will be a future PR

### CODE REVIEW

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/Condition.java [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+import javax.annotation.Nonnull;
+
+// TODO: Sub-types
+//  - Any (wildcard, useful where any execution of a 'MethodLocation' is flag-worthy)
+//  - Value matching for (Parameter N), (Field $NAME/$DESC)
+//    - Primitive math matching
+//    - Object null/not-null
+//    - String TextMatchMode matching
+//      - Also for char[] and other similar types (StringBuilder, StringBuffer, non-string CharSequence)
+//    - byte[] matching helpers
```
Comment (Col-E) [inline concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/Condition.java:14]: TODO: The condition sub-types need to be implemented still + the serializer/deserializer _(and this class annotated to use them)_

File: concoction-lib/src/test/java/info/mmpa/concoction/scan/model/DynamicMatchSerializationTests.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import static info.mmpa.concoction.util.TestSerialization.*;
+import static org.junit.jupiter.api.Assertions.*;
+
+
+/**
+ * Tests for {@link DynamicMatchingModel}, {@link DynamicMatchEntry} and {@link Condition} serialization.
+ */
+public class DynamicMatchSerializationTests {
+
+}
```
Comment (Col-E) [inline concoction-lib/src/test/java/info/mmpa/concoction/scan/model/DynamicMatchSerializationTests.java:27]: Tests need to be added once the `Condition` sub-types are created.

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/DynamicMatchEntrySerializer.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+		super(DynamicMatchEntry.class);
+	}
+
+	@Override
+	public void serialize(DynamicMatchEntry entry, JsonGenerator jgen, SerializerProvider provider) throws IOException {
+		if (entry instanceof SingleConditionCheckingDynamic) {
+			SingleConditionCheckingDynamic conditionCheck = (SingleConditionCheckingDynamic) entry;
+			jgen.writeStartObject();
+			jgen.writeObjectField("location", conditionCheck.getLocation());
+			if (conditionCheck.getWhen() != When.ENTRY) {
```
Comment (xyzeva) [inline concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/DynamicMatchEntrySerializer.java:28]: Nit: This comment can be moved up to above the if statement

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/MethodLocation.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+		this.methodName = methodName;
+		this.methodDesc = methodDesc;
+		this.classMatchMode = classMatchMode;
+		this.methodNameMatchMode = methodNameMatchMode;
+		this.methodDescMatchMode = methodDescMatchMode;
+	}
+
+	/**
+	 * @param frame
+	 * 		Frame to check.
```
Comment (xyzeva) [inline concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/MethodLocation.java:53]: Space in javadoc `@param`

File: concoction-lib/src/test/java/info/mmpa/concoction/scan/model/InstructionMatchSerializationTests.java [hunk 5]
```diff
@@ -16,7 +17,7 @@
 import static org.junit.jupiter.api.Assertions.*;
 
 /**
- * Tests for {@link InstructionMatchEntry} serialization.
+ * Tests for {@link InstructionsMatchingModel} and {@link InstructionMatchEntry} serialization.
  */
 public class InstructionMatchSerializationTests {
```
Comment (xyzeva) [inline concoction-lib/src/test/java/info/mmpa/concoction/scan/model/InstructionMatchSerializationTests.java:22]: Nit: Shouldn't these changes be in a different PR?

**Review by xyzeva — CHANGES_REQUESTED** [review]
Overall good, few changes needed here and there

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/MethodLocation.java (hunk 4 above)
Comment (Col-E) [reply concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/MethodLocation.java:53]: ?

File: concoction-lib/src/test/java/info/mmpa/concoction/scan/model/InstructionMatchSerializationTests.java (hunk 5 above)
Comment (Col-E) [reply concoction-lib/src/test/java/info/mmpa/concoction/scan/model/InstructionMatchSerializationTests.java:22]: I was poking around at all the other model stuff anyways

File: concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/MethodLocation.java (hunk 4 above)
Comment (xyzeva) [reply concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/MethodLocation.java:53]: On my screen, theres a space between
```
@param name
```
and
`Frame to check.`

Comment (Col-E) [reply concoction-lib/src/main/java/info/mmpa/concoction/scan/model/dynamic/entry/MethodLocation.java:53]: This is how everything else in the project is formatted. Its consistent, and IMO vastly improves legibility in non-rendered Javadoc.
